crates/nu-cmd-base/src/wrap_call.rs

```diff
@@ -0,0 +1,101 @@
+use nu_engine::CallExt;
+use nu_protocol::{
+    DeclId, FromValue, ShellError, Span,
+    engine::{Call, EngineState, Stack, StateWorkingSet},
+};
+
+/// A helper utility to aid in implementing commands which have the same behavior for `run` and `run_const`.
+///
+/// Only supports functions in [`Call`] and [`CallExt`] which have a `const` suffix.
+///
+/// To use, the actual command logic should be moved to a function. Then, `eval` and `eval_const` can be implemented like this:
+/// ```rust
+/// # use nu_engine::command_prelude::*;
+/// # use nu_cmd_base::WrapCall;
+/// # fn do_command_logic(call: WrapCall) -> Result<PipelineData, ShellError> { Ok(PipelineData::Empty) }
+///
+/// # struct Command {}
+/// # impl Command {
+/// fn run(&self, engine_state: &EngineState, stack: &mut Stack, call: &Call) -> Result<PipelineData, ShellError> {
+///     let call = WrapCall::Eval(engine_state, stack, call);
+///     do_command_logic(call)
+/// }
+///
+/// fn run_const(&self, working_set: &StateWorkingSet, call: &Call) -> Result<PipelineData, ShellError> {
+///     let call = WrapCall::ConstEval(working_set, call);
+///     do_command_logic(call)
+/// }
+/// # }
+/// ```
+///
+/// Then, the typical [`Call`] and [`CallExt`] operations can be called using destructuring:
+///
+/// ```rust
+/// # use nu_engine::command_prelude::*;
+/// # use nu_cmd_base::WrapCall;
+/// # let call = WrapCall::Eval(&EngineState::new(), &mut Stack::new(), &Call::new(Span::unknown()));
+/// # fn do_command_logic(call: WrapCall) -> Result<(), ShellError> {
+/// let (call, required): (_, String) = call.req(0)?;
+/// let (call, flag): (_, Option<i64>) = call.get_flag("number")?;
+/// # Ok(())
+/// # }
+/// ```
+///
+/// A new `WrapCall` instance has to be returned after each function to ensure
+/// that there is only ever one copy of mutable [`Stack`] reference.
+pub enum WrapCall<'a> {
```

Comment:
That type feels odd to me, I think we can come up with a better way of implementing this. But I don't see this as a blocker rn.